pkg/pillar/containerd/containerd.go

```diff
@@ -229,22 +178,48 @@ func CtrPrepare(containerPath string, ociFilename string) error {
 	log.Infof("mountPoints %+v execpath %+v cmd %+v workdir %+v env %+v",
 		mountpoints, execpath, cmd, workdir, unProcessedEnv)
 
-	if err != nil {
-		return fmt.Errorf("ctrPrepare: unable to get image: %v config: %v", ctrdImage.Name(), err)
-	}
-	// containerd.NewImageWithPlatform(client, i, platforms.Only(platform))
+	// unpack the rootfs Image if needed
 	unpacked, err := ctrdImage.IsUnpacked(ctrdCtx, defaultSnapshotter)
 	if err != nil {
-		return fmt.Errorf("ctrPrepare: unable to get image metadata: %v config: %v", ctrdImage.Name(), err)
+		return fmt.Errorf("snapshotPrepare: unable to get image metadata: %v config: %v", ctrdImage.Name(), err)
 	}
 	if !unpacked {
 		if err := ctrdImage.Unpack(ctrdCtx, defaultSnapshotter); err != nil {
-			return fmt.Errorf("ctrPrepare: unable to unpack image: %v config: %v", ctrdImage.Name(), err)
+			return fmt.Errorf("snapshotPrepare: unable to unpack image: %v config: %v", ctrdImage.Name(), err)
 		}
 	}
-	if err = ctrCreate(filepath.Base(containerPath), ctrdImage); err != nil {
-		return fmt.Errorf("ctrPrepare: failed to create container %s, error: %v", containerPath, err.Error())
+
+	// use rootfs unpacked image to create a writable snapshot with default snapshotter
+	diffIDs, err := ctrdImage.RootFS(ctrdCtx)
+	if err != nil {
+		log.Errorf("Could not load rootfs of image: %v. %v", ctrdImage.Name(), err)
+		return fmt.Errorf("snapshotPrepare: Could not load rootfs of image: %v. %v", ctrdImage.Name(), err)
+	}
+
+	snapshotter := CtrdClient.SnapshotService(defaultSnapshotter)
+	parent := identity.ChainID(diffIDs).String()
+	snapshot := getSnapshotName(containerID)
+	labels := map[string]string{"containerd.io/gc.root": time.Now().UTC().Format(time.RFC3339)}
+	if mounts, err := snapshotter.Prepare(ctrdCtx, snapshot, parent, snapshots.WithLabels(labels)); err != nil {
+		log.Errorf("Could not create a snapshot from: %s. %v", parent, err)
+		return fmt.Errorf("snapshotPrepare: Could not create a snapshot from: %s. %v", parent, err)
+	} else {
+		log.Infof("snapshotPrepare: preared a snapshot for %v with the following mounts: %v", snapshot, mounts)
+	}
+
+	// retrieve image metadata to be passed into createBundle
+	imageConfigJSON, err := getImageInfoJSON(ctrdCtx, ctrdImage)
+	if err != nil {
+		log.Errorf("Could not build json of image: %v. %v", ctrdImage.Name(), err.Error())
+		return fmt.Errorf("snapshotPrepare: Could not build json of image: %v. %v", ctrdImage.Name(), err.Error())
 	}
+
+	err = createBundle(containerID, getSnapshotName(containerID), imageConfigJSON)
+	if err != nil {
+		log.Errorf("Could not build rootfs of container: %v. %v", containerID, err.Error())
+		return fmt.Errorf("snapshotPrepare: Could not build rootfs of container: %v. %v", containerID, err.Error())
+	}
+
 	return nil
 }
 
```

Comment:
Can we remove `loadContainer(containerID string)` method?

Reply:
Yes we can, but since I'll be adding it immediately in the next PR -- I figured I'll keep it (and refactor it).